File: src/main/java/com/qtech/bubbles/entity/player/PlayerEntity.java
package com.qtech.bubbles.entity.player;

import com.google.common.annotations.Beta;
import com.qtech.bubbles.BubbleBlaster;
import com.qtech.bubbles.LoadedGame;
import com.qtech.bubbles.common.AttributeMap;
import com.qtech.bubbles.common.gametype.AbstractGameType;
import com.qtech.bubbles.entity.*;
import com.qtech.bubbles.entity.ammo.AmmoType;
import com.qtech.bubbles.entity.attribute.Attribute;
import com.qtech.bubbles.entity.damage.DamageSource;
import com.qtech.bubbles.entity.player.ability.AbilityContainer;
import com.qtech.bubbles.environment.Environment;
import com.qtech.bubbles.event.CollisionEvent;
import com.qtech.bubbles.event._common.SubscribeEvent;
import com.qtech.bubbles.event.input.KeyboardEvent;
import com.qtech.bubbles.event.input.XInputEvent;
import com.qtech.bubbles.event.type.KeyEventType;
import com.qtech.bubbles.init.AmmoTypes;
import com.qtech.bubbles.init.Entities;
import com.qtech.bubbles.item.inventory.PlayerInventory;
import com.qtech.bubbles.screen.EnvLoadScreen;
import com.qtech.bubbles.screen.SaveLoadingScreen;
import com.qtech.bubbles.screen.Screen;
import com.qtech.bubbles.util.Util;
import com.qtech.bubbles.util.helpers.MathHelper;
import org.apache.batik.ext.awt.geom.Polygon2D;
import org.apache.commons.lang3.SystemUtils;
import org.bson.BsonDocument;
import org.bson.BsonDouble;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.awt.*;
import java.awt.event.KeyEvent;
import java.awt.geom.*;
import java.util.Objects;

import static com.qtech.bubbles.BubbleBlaster.TPS;

@SuppressWarnings({"RedundantSuppression", "unused"})
public final class PlayerEntity extends DamageableEntity {

    // These are the vertex coordinates:
    //
    //          |- Middle point
    //          |
    // 9876543210123456789
    // --------------------,
    // ....#*********#.....| -8
    // ....*.........*.....| -7
    // ....#***#.#***#.....| -6
    // ........*.*.........| -5
    // ..#*****#.#******#..| -4
    // .*..................| -3
    // #..................#| -2
    // *..................*| -1
    // *..................*| 0  // Middle point.
    // *..................*| 1
    // #..................#| 2
    // .*................*.| 3
    // ..#..............#..| 4

    private final Area shipShape;
    private final Area arrowShape;

    // Types
    private AmmoType currentAmmo = AmmoTypes.BASIC.get();

    // Motion (Arrow Keys).
    private boolean forward = false;
    private boolean backward = false;
    private boolean left = false;
    private boolean right = false;

    // Speed.
    private final double rotationSpeed = 16f;

    // Delta velocity.
    private double velDelta;

    // Motion (XInput).
    private float xInputX;
    private float xInputY;

    // Rotation
    private double rotation = 0;

    // Normal values/
    private double score = 0d;
    private int level = 1;

    // Modifiers.
    private long abilityEnergy;

    // Ability
    private final AbilityContainer abilityContainer = new AbilityContainer(this);

    private double accelerateX = 0.0d;
    private double accelerateY = 0.0d;
    private final PlayerInventory inventory = new PlayerInventory(this);

    /**
     * <h1>Player entity.</h1>
     * The player is controlled be the keyboard and is one of the important features of the game. (Almost any game).
     *
     * @see DamageableEntity
     */
    public PlayerEntity(AbstractGameType gameType) {
        super(Entities.PLAYER.get(), gameType);

        this.addCollidable(Entities.BUBBLE.get());

        // Ship shape.
        Ellipse2D shipShape1 = new Ellipse2D.Double(-20, -20, 40, 40);
        Ellipse2D shipShape2 = new Ellipse2D.Double(-18, -18, 36, 36);
        Ellipse2D shipShape3 = new Ellipse2D.Double(-16, -16, 32, 32);
        Ellipse2D shipShape4 = new Ellipse2D.Double(-14, -14, 28, 28);
        this.shipShape = new Area(shipShape1);

        // Arrow shape.
        Polygon2D arrowShape1 = new Polygon2D(new int[]{-5, -10, 15, -10}, new int[]{0, -10, 0, 10}, 4);
        this.arrowShape = new Area(arrowShape1);

        // Velocity.
        this.velX = 0;
        this.velY = 0;

        // Set attributes.
        this.attributes.setBase(Attribute.DEFENSE, 1f);
        this.attributes.setBase(Attribute.ATTACK, 0.75f);
        this.attributes.setBase(Attribute.MAX_DAMAGE, 100f);
        this.attributes.setBase(Attribute.SPEED, 16f);

        // Health
        this.speed = 16f;
        this.damageValue = 100f;
    }

    /**
     * @param s the message.
     */
    public void sendMessage(String s) {
        @Nullable Screen currentScene = Util.getSceneManager().getCurrentScreen();
        LoadedGame loadedGame = BubbleBlaster.getInstance().getLoadedGame();
        if (loadedGame != null) {
            loadedGame.triggerMessage(s);
        }
    }

    /**
     * <h1>Prepare spawn</h1>
     * Called when the entity was spawned.
     *
     * @param spawnData the data to spawn with.
     */
    @Override
    public void prepareSpawn(EntitySpawnData spawnData) {
        super.prepareSpawn(spawnData);
        @Nullable Screen currentScene = Objects.requireNonNull(Util.getSceneManager()).getCurrentScreen();
        if ((currentScene == null && BubbleBlaster.getInstance().isGameLoaded()) ||
                currentScene instanceof SaveLoadingScreen ||
                currentScene instanceof EnvLoadScreen) {
            Rectangle2D gameBounds = gameType.getGameBounds();
            this.x = (float) MathHelper.clamp(x, gameBounds.getMinX(), gameBounds.getMaxX());
            this.y = (float) MathHelper.clamp(y, gameBounds.getMinY(), gameBounds.getMaxY());
            bindEvents();
        }
    }

    @Override
    protected void bindEvents() {
        BubbleBlaster.getEventBus().register(this);
        this.areEventsBinded = true;
    }

    @Override
    protected void unbindEvents() {
        BubbleBlaster.getEventBus().unregister(this);
        this.areEventsBinded = false;
    }

    @Override
    protected boolean areEventsBound() {
        return this.areEventsBinded;
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    //     Shape     //
    ///////////////////
    @Override
    public Area getShape() {
        return rotateToArea();
    }

    @NotNull
    private Area rotateToArea() {
        return rotateToArea0(shipShape);
    }

    @NotNull
    private Area rotateToArea0(Area shipShape) {
        AffineTransform transform = new AffineTransform();
        transform.rotate(Math.toRadians(rotation), shipShape.getBounds().getCenterX(), shipShape.getBounds().getCenterY());

        Area area = new Area(shipShape);
        area.transform(transform);

        transform = new AffineTransform();
        transform.scale(scale, scale);
        transform.translate(x, y);

        area.transform(transform);

        return area;
    }

    /**
     * @return the x acceleration.
     */
    public double getAccelerateX() {
        return accelerateX;
    }

    /**
     * @return the y acceleration.
     */
    public double getAccelerateY() {
        return accelerateY;
    }

    /**
     * @param accelerateX the x acceleration to set.
     */
    public void setAccelerateX(double accelerateX) {
        this.accelerateX = accelerateX;
    }

    /**
     * @param accelerateY the y acceleration to set.
     */
    public void setAccelerateY(double accelerateY) {
        this.accelerateY = accelerateY;
    }

    /**
     * @return the shape of the ship.
     */
    public Area getShipArea() {
        return rotateToArea();
    }

    /**
     * @return the arrow shape of the ship.
     */
    public Area getArrowArea() {
        return rotateToArea0(arrowShape);
    }

    /**
     * @return the center position.
     */
    @SuppressWarnings("unused")
    public Point getCenter() {
        return new Point((int) getBounds().getCenterX(), (int) getBounds().getCenterY());
    }

    /**
     * Tick the player.
     *
     * @param environment the game-type where the entity is from.
     */
    @Override
    @SubscribeEvent
    public void tick(Environment environment) {
        super.tick(environment);

        ////////////////////////////////////////////////////////////////////////////////////////////////////////////////
        // Spawn and load checks //
        ///////////////////////////

        LoadedGame loadedGame = BubbleBlaster.getInstance().getLoadedGame();

        if (loadedGame == null) {
            return;
        }

        if (!isSpawned()) return;

        ////////////////////////////////////////////////////////////////////////////////////////////////////////////////
        // Player component ticking //
        //////////////////////////////

        this.abilityContainer.onEntityTick();
        this.inventory.tick();

        ////////////////////////////////////////////////////////////////////////////////////////////////////////////////
        // Player motion. //
        ////////////////////

        this.accelerateX = accelerateX / ((double)TPS / 20d) / 1.1;
        this.accelerateY = accelerateY / ((double)TPS / 20d) / 1.1;

        double tempVelMotSpeed = 0.0f;
        double tempVelRotSpeed = 0.0f;

        // Check each direction, to create velocity
        if (this.forward) tempVelMotSpeed += this.speed;
        if (this.backward) tempVelMotSpeed -= this.speed;
        if (this.left) tempVelRotSpeed -= this.rotationSpeed;
        if (this.right) tempVelRotSpeed += this.rotationSpeed;
        if (this.xInputY != 0) tempVelMotSpeed = this.xInputY * this.speed;
        if (this.xInputX != 0) tempVelRotSpeed = this.xInputX * this.rotationSpeed;

        // Update rotation
        double delta = 1d / TPS;

        // Update X, and Y.
        if (isMotionEnabled()) {
            this.rotation += (tempVelRotSpeed) / TPS * 20;
        }

        // Calculate Velocity X and Y.
        double angelRadians = Math.toRadians(this.rotation);
        double tempVelX = Math.cos(angelRadians) * tempVelMotSpeed;
        double tempVelY = Math.sin(angelRadians) * tempVelMotSpeed;

//        System.out.println("Acc_X[0] = " + accelerateX);
//        System.out.println("Acc_Y[0] = " + accelerateY);

        if (isMotionEnabled()) {
            this.accelerateX += tempVelX / ((double) TPS);
            this.accelerateY += tempVelY / ((double) TPS);
        }

//        System.out.println("Acc_X[1] = " + accelerateX);
//        System.out.println("Acc_Y[1] = " + accelerateY);

        // Update X, and Y.
        this.x += (this.accelerateX) + this.velX / ((double) TPS * 1);
        this.y += (this.accelerateY) + this.velY / ((double) TPS * 1);

        // Velocity.
        if (this.velX > 0) {
            if (this.velX + this.velDelta < 0) {
                this.velX = 0;
            } else {
                this.velX += this.velDelta;
            }
        } else if (this.velX < 0) {
            if (this.velX + this.velDelta > 0) {
                this.velX = 0;
            } else {
                this.velX -= this.velDelta;
            }
        }

        if (this.velY > 0) {
            if (this.velY + velDelta < 0) {
                this.velY = 0;
            } else {
                this.velY += this.velDelta;
            }
        } else if (this.velX < 0) {
            if (this.velY + this.velDelta > 0) {
                this.velY = 0;
            } else {
                this.velY -= this.velDelta;
            }
        }

        ////////////////////////////////////////////////////////////////////////////////////////////////////////////////
        // Collision detection. //
        //////////////////////////

        // Game border.
        Rectangle2D gameBounds = loadedGame.getGameType().getGameBounds();
        this.x = (float) MathHelper.clamp(this.x, gameBounds.getMinX() - this.getBounds().getWidth(), gameBounds.getMaxX() + this.getBounds().getWidth());
        this.y = (float) MathHelper.clamp(this.y, gameBounds.getMinY() - this.getBounds().getHeight(), gameBounds.getMaxY() + this.getBounds().getHeight());
    }

    @Override
    public void damage(double value, DamageSource source) {
        float defense = attributes.getBase(Attribute.DEFENSE);
        if (defense <= 0.0d) {
            this.destroy();
            return;
        }

        // Deal damage to the player.
        this.damageValue -= value / defense;

        // Check health.
        this.checkDamage();

        // Check if source has attack modifier.
        if (value > 0.0d) {

            // Check if window is not focused.
            if (!BubbleBlaster.getInstance().getFrame().isFocused()) {
                if (SystemUtils.IS_JAVA_9) {
                    // Let the taskbar icon flash. (Java 9+)
                    @SuppressWarnings("Since15") Taskbar taskbar = Taskbar.getTaskbar();
                    try {
                        //noinspection Since15
                        taskbar.requestWindowUserAttention(BubbleBlaster.getInstance().getFrame());
                    } catch (UnsupportedOperationException ignored) {

                    }
                }
            }
        }
    }

    @SubscribeEvent
    public void onCollision(CollisionEvent evt) {
        if (evt.getSource() != this) return;

        Entity targetEntity = evt.getTarget();

        // Entity bubble
        if (targetEntity instanceof GiantBubbleEntity bubble) {

            // Modifiers
            double scoreMod = bubble.getAttributeMap().getBase(Attribute.SCORE_MULTIPLIER);
            double attackMod = bubble.getAttributeMap().getBase(Attribute.ATTACK);
            double defenseMod = bubble.getAttributeMap().getBase(Attribute.DEFENSE);

            // Attributes
            double radius = bubble.getRadius();
            double speed = bubble.getSpeed();

            // Calculate score value.
            double visibleValue = radius * speed;
            double nonVisibleValue = attackMod * defenseMod;
            double scoreValue = ((visibleValue * (nonVisibleValue + 1)) * scoreMod * attributes.getBase(Attribute.SCORE_MULTIPLIER)) * evt.getDeltaTime() / 8;

            // Add score.
            addScore(scoreValue);
        } else
            // Entity bubble
            if (targetEntity instanceof BubbleEntity bubble) {
                // Modifiers
                AttributeMap attributeMap = bubble.getAttributeMap();
                double scoreMultiplier = attributeMap.getBase(Attribute.SCORE_MULTIPLIER);
                double attack = attributeMap.getBase(Attribute.ATTACK);  // Maybe used.
                double defense = attributeMap.getBase(Attribute.DEFENSE);  // Maybe used.

                // Attributes
                double radius = bubble.getRadius();
                double speed = bubble.getSpeed();

            // Calculate score value.
            double visibleValue = radius * speed;
            double nonVisibleValue = attack * defense;
            double scoreValue = ((visibleValue * (nonVisibleValue + 1)) * scoreMultiplier * scoreMultiplier) * evt.getDeltaTime() / 8;

            // Add score.
            addScore(scoreValue);
        }
    }

    @SubscribeEvent
    public void onKeyboardEvent(KeyboardEvent evt) {
        if (evt != null) {
            KeyEvent e = evt.getParentEvent();

            if (evt.getType() == KeyEventType.PRESS) {
                if (e.getKeyCode() == KeyEvent.VK_UP) forward = true;
                if (e.getKeyCode() == KeyEvent.VK_DOWN) backward = true;
                if (e.getKeyCode() == KeyEvent.VK_LEFT) left = true;
                if (e.getKeyCode() == KeyEvent.VK_RIGHT) right = true;
            }

            if (evt.getType() == KeyEventType.RELEASE) {
                if (e.getKeyCode() == KeyEvent.VK_UP) forward = false;
                if (e.getKeyCode() == KeyEvent.VK_DOWN) backward = false;
                if (e.getKeyCode() == KeyEvent.VK_LEFT) left = false;
                if (e.getKeyCode() == KeyEvent.VK_RIGHT) right = false;
            }
        }
    }

    @SubscribeEvent
    public void onXInputEvent(XInputEvent evt) {
        System.out.println(evt);

        if (evt != null) {
            if (forward || backward || left || right) return;

            this.xInputX = evt.getRightStickX();
            this.xInputY = evt.getLeftStickY();
        }
    }

    @Override
    public void renderEntity(Graphics2D gg) {
        if (!isSpawned()) return;

        ////////////////////////////////////////////////////////////////////////////////////////////////////////////////
        // Fill the ship with the correct color. //
        ///////////////////////////////////////////
        gg.setColor(Color.red);
        gg.fill(getShipArea());

        ////////////////////////////////////////////////////////////////////////////////////////////////////////////////
        // Fill the arrow with the correct color. //
        ////////////////////////////////////////////
        gg.setColor(Color.white);
        gg.fill(getArrowArea());
    }

    /////////////////////////////////////////////////////////////////////////////////////
    //     Rotate Delta     //
    //////////////////////////
    @SuppressWarnings("unused")
    public void rotateDelta(int deltaRotation) {
        this.rotation += deltaRotation;
    }

    /**
     * <h1>Trigger a Reflection</h1>
     * Triggers a reflection, there are some problems with the velocity.
     * That's why it's currently in beta.
     *
     * @param velocity the amount velocity for bounce.
     * @param delta    the delta change.
     */
    @Beta
    public void triggerReflect(Point2D velocity, double delta) {
        this.velDelta = delta;
        this.velX = velocity.getX();
        this.velY = velocity.getY();
    }

    /**
     * <h1>Delete method</h1>
     * <i><b>WARNING: </b><u>This is an unsafe method, use {@link AbstractGameType#triggerGameOver()} instead.</u></i>
     *
     * @see AbstractGameType#triggerGameOver()
     */
    @Override
    public void delete() {
        this.unbindEvents();
    }

    /**
     * <h1>Checks health</h1>
     * Checks health, if health is zero or less, it will trigger game-over.
     */
    @Override
    public void checkDamage() {
        if (damageValue <= 0)
            if (!getGameType().isGameOver())
                getGameType().triggerGameOver();
    }

    ///////////////////////////////////////////////////////////////////////////////////////////////////////////////
    //     Entity Data     //
    /////////////////////////
    @Override
    public @NotNull BsonDocument getState() {
        BsonDocument document = super.getState();

        document.put("score", new BsonDouble(score));
        document.put("rotation", new BsonDouble(rotation));
        return document;
    }

    @Override
    public void setState(BsonDocument state) {
        super.getState();

        this.score = state.getDouble("score").getValue();
        this.rotation = state.getDouble("rotation").getValue();
    }

    ///////////////////////////////////////////////////////////////////////////////////////////////////////////////
    //     Speed     //
    ///////////////////
    public double getRotationSpeed() {
        return rotationSpeed;
    }

    ///////////////////////////////////////////////////////////////////////////////////////////////////////////////
    //     Rotation     //
    //////////////////////
    public double getRotation() {
        return rotation;
    }

    public void setRotation(double rotation) {
        this.rotation = rotation;
    }

    ///////////////////////////////////////////////////////////////////////////////////////////////////////////////
    //     Values     //
    ////////////////////

    // Score
    public double getScore() {
        return score;
    }

    public void addScore(double value) {
        score += value;
    }

    public void subtractScore(long value) {
        score -= value;
    }

    public void setScore(double value) {
        score = value;
    }

    // Level
    public int getLevel() {
        return level;
    }

    public void levelUp() {
        level++;
    }

    public void levelDown() {
        setLevel(getLevel() - 1);
    }

    public void setLevel(int level) {
        this.level = Math.max(level, 1);
    }

    ///////////////////////////////////////////////////////////////////////////////////////////////////////////////
    //     Modifiers     //
    ///////////////////////
    /**
     * Get the ability container of the entity.
     *
     * @return the requested {@link AbilityContainer}.
     * @see AbilityContainer
     */
    public AbilityContainer getAbilityContainer() {
        return abilityContainer;
    }

    /**
     * Get current ammo type.
     *
     * @return the current {@link AmmoType ammo type}.
     * @see AmmoType
     * @see #setCurrentAmmo(AmmoType)
     */
    public AmmoType getCurrentAmmo() {
        return currentAmmo;
    }

    /**
     * Set current ammo type.
     *
     * @param currentAmmo the ammo to set.
     * @see AmmoType
     * @see #getCurrentAmmo()
     */
    public void setCurrentAmmo(AmmoType currentAmmo) {
        this.currentAmmo = currentAmmo;
    }

    /**
     * Set current ship scale.
     *
     * @param scale the scale to set.
     * @see #getScale()
     */
    public void setScale(double scale) {
        this.scale = scale;
    }

    void forward(boolean bool) {
        this.forward = bool;
    }

    void backward(boolean bool) {
        this.backward = bool;
    }

    void left(boolean bool) {
        this.left = bool;
    }

    void right(boolean bool) {
        this.right = bool;
    }
}

File: src/main/kotlin/com/qtech/bubbles/entity/player/PlayerEntity.kt
@file:Suppress("unused")

package com.qtech.bubbles.entity.player

import com.google.common.annotations.Beta
import com.qtech.bubbles.BubbleBlaster
import com.qtech.bubbles.common.ability.AbilityContainer
import com.qtech.bubbles.common.ammo.AmmoType
import com.qtech.bubbles.common.entity.*
import com.qtech.bubbles.common.gametype.AbstractGameMode
import com.qtech.bubbles.entity.DamageableEntity
import com.qtech.bubbles.environment.Environment
import com.qtech.bubbles.event.*
import com.qtech.bubbles.event.type.KeyEventType
import com.qtech.bubbles.graphics.GraphicsProcessor
import com.qtech.bubbles.init.AmmoTypes
import com.qtech.bubbles.init.Entities
import com.qtech.bubbles.item.inventory.PlayerInventory
import com.qtech.bubbles.screen.EnvLoadScreen
import com.qtech.bubbles.util.Util
import com.qtech.bubbles.util.helpers.MathHelper
import net.querz.nbt.tag.CompoundTag
import org.apache.batik.ext.awt.geom.Polygon2D
import org.apache.commons.lang3.SystemUtils
import java.awt.Color
import java.awt.Point
import java.awt.Taskbar
import java.awt.event.KeyEvent
import java.awt.geom.AffineTransform
import java.awt.geom.Area
import java.awt.geom.Ellipse2D
import java.awt.geom.Point2D
import java.util.*
import kotlin.math.cos
import kotlin.math.sin

class PlayerEntity(gameMode: AbstractGameMode?) : DamageableEntity(Entities.PLAYER.get(), gameMode!!) {
    // These are the vertex coordinates:
    //
    //          |- Middle point
    //          |
    // 9876543210123456789
    // --------------------,
    // ....#*********#.....| -8
    // ....*.........*.....| -7
    // ....#***#.#***#.....| -6
    // ........*.*.........| -5
    // ..#*****#.#******#..| -4
    // .*..................| -3
    // #..................#| -2
    // *..................*| -1
    // *..................*| 0  // Middle point.
    // *..................*| 1
    // #..................#| 2
    // .*................*.| 3
    // ..#..............#..| 4
    private val shipShape: Area
    private val arrowShape: Area

    // Ammo
    var currentAmmo: AmmoType = AmmoTypes.BASIC.get()

    // Rotation
    public override var rotation = 0.0

    // Normal values.
    var score = 0.0
    var level = 1
        get() = field.coerceAtLeast(1).also { field = it }
        set(value) {
            field = value.coerceAtLeast(1)
        }

    // Modifiers.
    private val abilityEnergy: Long = 0

    // Ability
    val abilityContainer = AbilityContainer(this)

    // Acceleration
    var accelerateX = 0.0
    var accelerateY = 0.0

    // Motion (Arrow Keys).
    private var forward = false
    private var backward = false
    private var left = false
    private var right = false

    // Speed.
    val rotationSpeed = 16f

    // Delta velocity.
    private var velDelta = 0.0

    // Motion (XInput).
    private var xInputX = 0f
    private var xInputY = 0f

    //
    val inventory = PlayerInventory(this)

    var scoreMultiplier: Float
        get() = attributeMap[Attribute.SCORE_MULTIPLIER]
        set(value) {
            attributeMap[Attribute.SCORE_MULTIPLIER] = value
        }

    var baseScoreMultiplier: Float
        get() = bases[Attribute.SCORE_MULTIPLIER]
        private set(value) {
            bases[Attribute.SCORE_MULTIPLIER] = value
        }

    var luck: Float
        get() = attributeMap[Attribute.LUCK]
        set(value) {
            attributeMap[Attribute.LUCK] = value
        }

    var baseLuck: Float
        get() = bases[Attribute.LUCK]
        private set(value) {
            bases[Attribute.LUCK] = value
        }

    /**
     * @param s the message.
     */
    fun sendMessage(s: String) {
        val loadedGame = BubbleBlaster.instance.loadedGame
        loadedGame?.triggerMessage(s)
    }

    /**
     * <h1>Prepare spawn</h1>
     * Called when the entity was spawned.
     *
     * @param spawnData the data to spawn with.
     */
    override fun prepareSpawn(spawnData: EntitySpawnData) {
        super.prepareSpawn(spawnData)
        val currentScene = Objects.requireNonNull(Util.sceneManager).currentScreen
        if (currentScene == null && BubbleBlaster.instance.isGameLoaded ||
            currentScene is EnvLoadScreen
        ) {
            val gameBounds = gameMode.gameBounds
            x = MathHelper.clamp(x.toDouble(), gameBounds!!.minX, gameBounds.maxX).toFloat()
            y = MathHelper.clamp(y.toDouble(), gameBounds.minY, gameBounds.maxY).toFloat()
            bindEvents()
        }
    }

    override fun bindEvents() {
        BubbleBlaster.eventBus.register(this)
        areEventsBound = true
    }

    override fun unbindEvents() {
        BubbleBlaster.eventBus.unregister(this)
        areEventsBound = false
    }

    override fun areEventsBound(): Boolean {
        return areEventsBound
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    //     Shape     //
    ///////////////////
    override val shape: Area
        get() {
            var transform = AffineTransform()
            transform.rotate(Math.toRadians(rotation), shipShape.bounds.centerX, shipShape.bounds.centerY)
            val area = Area(shipShape)
            area.transform(transform)
            transform = AffineTransform()
            transform.scale(scale, scale)
            transform.translate(x.toDouble(), y.toDouble())
            area.transform(transform)
            return area
        }

    /**
     * @return the shape of the ship.
     */
    val shipArea: Area
        get() {
            var transform = AffineTransform()
            transform.rotate(Math.toRadians(rotation), shipShape.bounds.centerX, shipShape.bounds.centerY)
            val area = Area(shipShape)
            area.transform(transform)
            transform = AffineTransform()
            transform.scale(scale, scale)
            transform.translate(x.toDouble(), y.toDouble())
            area.transform(transform)
            return area
        }

    /**
     * @return the arrow shape of the ship.
     */
    val arrowArea: Area
        get() {
            var transform = AffineTransform()
            transform.rotate(Math.toRadians(rotation), shipShape.bounds.centerX, shipShape.bounds.centerY)
            val area = Area(arrowShape)
            area.transform(transform)
            transform = AffineTransform()
            transform.scale(scale, scale)
            transform.translate(x.toDouble(), y.toDouble())
            area.transform(transform)
            return area
        }

    /**
     * @return the center position.
     */
    val center: Point
        get() = Point(bounds!!.centerX.toInt(), bounds!!.centerY.toInt())

    /**
     * Tick the player.
     *
     * @param environment the game-type where the entity is from.
     */
    @SubscribeEvent
    override fun tick(environment: Environment?) {
        // Super call
        super.tick(environment)

        // Spawned and loaded checks
        val loadedGame = BubbleBlaster.instance.loadedGame ?: return
        if (!isSpawned) return

        // Component ticking
        abilityContainer.onEntityTick()
        inventory.tick()

        // Decelerate
        accelerateX = accelerateX / (BubbleBlaster.TPS.toDouble() / 20.0) / 1.05
        accelerateY = accelerateY / (BubbleBlaster.TPS.toDouble() / 20.0) / 1.05
        var tempVelMotSpeed = 0f
        var tempVelRotSpeed = 0f

        // Check each direction, to create velocity
        if (forward) tempVelMotSpeed += speed
        if (backward) tempVelMotSpeed -= speed
        if (left) tempVelRotSpeed -= rotationSpeed
        if (right) tempVelRotSpeed += rotationSpeed
        if (xInputY != 0f) tempVelMotSpeed = xInputY * speed
        if (xInputX != 0f) tempVelRotSpeed = xInputX * rotationSpeed

        // Update X, and Y.
        if (isMotionEnabled) {
            rotation += tempVelRotSpeed / BubbleBlaster.TPS * 20
        }

        // Calculate Velocity X and Y.
        val angelRadians = Math.toRadians(rotation)
        val tempVelX = cos(angelRadians) * tempVelMotSpeed
        val tempVelY = sin(angelRadians) * tempVelMotSpeed

        if (isMotionEnabled) {
            accelerateX += tempVelX / BubbleBlaster.TPS.toDouble()
            accelerateY += tempVelY / BubbleBlaster.TPS.toDouble()
        }

        // Update X, and Y.
        x += (accelerateX + velocityX / (BubbleBlaster.TPS.toDouble() * 1)).toFloat()
        y += (accelerateY + velocityY / (BubbleBlaster.TPS.toDouble() * 1)).toFloat()

        // Velocity.
        when {
            velocityX > 0 -> {
                if (velocityX + velDelta < 0) {
                    velocityX = 0f
                } else {
                    velocityX += velDelta.toFloat()
                }
            }
            velocityX < 0 -> {
                if (velocityX + velDelta > 0) {
                    velocityX = 0f
                } else {
                    velocityX -= velDelta.toFloat()
                }
            }
        }
        when {
            velocityY > 0 -> {
                if (velocityY + velDelta < 0) {
                    velocityY = 0f
                } else {
                    velocityY += velDelta.toFloat()
                }
            }
            velocityX < 0 -> {
                if (velocityY + velDelta > 0) {
                    velocityY = 0f
                } else {
                    velocityY -= velDelta.toFloat()
                }
            }
        }

        ////////////////////////////////////////////////////////////////////////////////////////////////////////////////
        // Collision detection. //
        //////////////////////////
        val width = bounds!!.getWidth() / 2
        val height = bounds!!.getHeight() / 2

        // Game border.
        val gameBounds = loadedGame.gameMode.gameBounds
        x = MathHelper.clamp(x.toDouble(), gameBounds!!.minX + width, gameBounds.maxX - width).toFloat()
        y = MathHelper.clamp(y.toDouble(), gameBounds.minY + height, gameBounds.maxY - height).toFloat()
    }

    override fun damage(value: Double, source: EntityDamageSource?) {
        val defense = attributeMap[Attribute.DEFENSE]
        if (defense <= 0.0) {
            destroy()
            return
        }

        // Deal damage to the player.
        damageValue -= (value / defense).toFloat()

        // Check health.
        checkDamage()

        // Check if source has attack modifier.
        if (value > 0.0) {

            // Check if window is not focused.
            if (!BubbleBlaster.instance.frame.isFocused) {
                if (SystemUtils.IS_JAVA_9) {
                    // Let the taskbar icon flash. (Java 9+)
                    val taskbar = Taskbar.getTaskbar()
                    try {
                        taskbar.requestWindowUserAttention(BubbleBlaster.instance.frame)
                    } catch (ignored: UnsupportedOperationException) {
                    }
                }
            }
        }
    }

    @SubscribeEvent
    override fun onCollision(event: CollisionEvent) {

    }

    @SubscribeEvent
    fun onKeyboardEvent(evt: KeyboardEvent?) {
        if (evt != null) {
            val e = evt.parentEvent
            if (evt.type == KeyEventType.PRESS) {
                if (e.keyCode == KeyEvent.VK_UP) forward = true
                if (e.keyCode == KeyEvent.VK_DOWN) backward = true
                if (e.keyCode == KeyEvent.VK_LEFT) left = true
                if (e.keyCode == KeyEvent.VK_RIGHT) right = true
            }
            if (evt.type == KeyEventType.RELEASE) {
                if (e.keyCode == KeyEvent.VK_UP) forward = false
                if (e.keyCode == KeyEvent.VK_DOWN) backward = false
                if (e.keyCode == KeyEvent.VK_LEFT) left = false
                if (e.keyCode == KeyEvent.VK_RIGHT) right = false
            }
        }
    }

    @SubscribeEvent
    fun onXInputEvent(evt: XInputEvent?) {
        println(evt)
        if (evt != null) {
            if (forward || backward || left || right) return
            xInputX = evt.rightStickX
            xInputY = evt.leftStickY
        }
    }

    @Synchronized
    override fun renderEntity(gg: GraphicsProcessor?) {
        if (!isSpawned) return

        ////////////////////////////////////////////////////////////////////////////////////////////////////////////////
        // Fill the ship with the correct color. //
        ///////////////////////////////////////////
        gg!!.color(Color.red)
        gg.shapeF(shipArea)

        ////////////////////////////////////////////////////////////////////////////////////////////////////////////////
        // Fill the arrow with the correct color. //
        ////////////////////////////////////////////
        gg.color(Color.white)
        gg.shapeF(arrowArea)
    }

    /////////////////////////////////////////////////////////////////////////////////////
    //     Rotate Delta     //
    //////////////////////////
    fun rotateDelta(deltaRotation: Int) {
        rotation += deltaRotation.toDouble()
    }

    /**
     * <h1>Trigger a Reflection</h1>
     * Triggers a reflection, there are some problems with the velocity.
     * That's why it's currently in beta.
     *
     * @param velocity the amount velocity for bounce.
     * @param delta    the delta change.
     */
    @Beta
    fun triggerReflect(velocity: Point2D, delta: Double) {
        velDelta = delta
        velocityX = velocity.x.toFloat()
        velocityY = velocity.y.toFloat()
    }

    /**
     * <h1>Delete method</h1>
     * ***WARNING: **<u>This is an unsafe method, use [AbstractGameMode.triggerGameOver] instead.</u>*
     *
     * @see AbstractGameMode.triggerGameOver
     */
    override fun delete() {
        unbindEvents()
    }

    /**
     * <h1>Checks health</h1>
     * Checks health, if health is zero or less, it will trigger game-over.
     */
    public override fun checkDamage() {
        if (damageValue <= 0) if (!gameMode.isGameOver) gameMode.triggerGameOver()
    }

    ///////////////////////////////////////////////////////////////////////////////////////////////////////////////
    //     Entity Data     //
    /////////////////////////
    override fun getState(): CompoundTag {
        val document = super.getState()
        document.putDouble("score", score)
        document.putDouble("rotation", rotation)
        return document
    }

    override fun setState(state: CompoundTag) {
        super.getState()
        score = state.getDouble("score")
        rotation = state.getDouble("rotation")
    }

    fun addScore(value: Double) {
        score += value
    }

    fun subtractScore(value: Long) {
        score -= value.toDouble()
    }

    fun levelUp() {
        level++
    }

    fun levelDown() {
        level--
    }

    fun forward(bool: Boolean) {
        forward = bool
    }

    fun backward(bool: Boolean) {
        backward = bool
    }

    fun left(bool: Boolean) {
        left = bool
    }

    fun right(bool: Boolean) {
        right = bool
    }

    /**
     * <h1>Player entity.</h1>
     * The player is controlled be the keyboard and is one of the important features of the game. (Almost any game).
     *
     * @see DamageableEntity
     */
    init {
        this.addCollidable(Entities.BUBBLE.get())

        // Ship shape.
        val shipShape1: Ellipse2D = Ellipse2D.Double(-20.0, -20.0, 40.0, 40.0)
        shipShape = Area(shipShape1)

        // Arrow shape.
        val arrowShape1 = Polygon2D(intArrayOf(-5, -10, 15, -10), intArrayOf(0, -10, 0, 10), 4)
        arrowShape = Area(arrowShape1)

        // Velocity.
        velocityX = 0f
        velocityY = 0f

        // Set attributes.
        attributeMap[Attribute.DEFENSE] = 1f
        attributeMap[Attribute.ATTACK] = 10f
        attributeMap[Attribute.MAX_DAMAGE] = 100f
        attributeMap[Attribute.SPEED] = 16f

        // Health
        speed = 10f
        damageValue = 100f
    }
}